src/Interpreters/QueryMetricLog.cpp

```diff
@@ -0,0 +1,180 @@
+
+    for (size_t i = 0, end = ProfileEvents::end(); i < end; ++i)
+        columns[column_idx++]->insert(profile_events[i]);
+}
+
+void QueryMetricLog::shutdown()
+{
+    stopCollect();
+    SystemLog<QueryMetricLogElement>::shutdown();
+}
+
+void QueryMetricLog::stopCollect()
+{
+    std::lock_guard lock(queries_mutex);
+    for (auto & [query_id, status] : queries)
+        status.task->deactivate();
+}
+
+void QueryMetricLog::startQuery(const String & query_id, TimePoint query_start_time, UInt64 interval_milliseconds)
+{
+    QueryMetricLogStatus status;
+    status.interval_milliseconds = interval_milliseconds;
+    status.next_collect_time = query_start_time + std::chrono::milliseconds(interval_milliseconds);
+
+    const auto & profile_events = CurrentThread::getProfileEvents();
+    for (ProfileEvents::Event i = ProfileEvents::Event(0), end = ProfileEvents::end(); i < end; ++i)
+        status.last_profile_events[i] = profile_events[i].load(std::memory_order_relaxed);
+
+    auto context = getContext();
+    const auto & process_list = context->getProcessList();
+    status.task = context->getSchedulePool().createTask("QueryLog", [this, &process_list, query_id] {
+        auto current_time = std::chrono::system_clock::now();
+        const auto query_info = process_list.getQueryInfo(query_id, false, true, false);
+
+        /// The query info should always be found because whenever a query ends, finishQuery is
+        /// called and the query is removed from the list
+        if (!query_info)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Query info not found: {}", query_id);
+
+        auto elem = createLogMetricElement(query_id, query_info->profile_counters, current_time);
+        add(std::move(elem));
+    });
+
+    status.task->scheduleAfter(interval_milliseconds);
+
+    std::lock_guard lock(queries_mutex);
+    queries.emplace(query_id, std::move(status));
+}
+
+void QueryMetricLog::finishQuery(const String & query_id)
+{
+    std::lock_guard lock(queries_mutex);
+    auto it = queries.find(query_id);
+
+    /// finishQuery may be called from logExceptionBeforeStart when the query has not even started
+    /// yet, so its corresponding startQuery is never called.
+    if (it == queries.end())
+        return;
+
+    it->second.task->deactivate();
+    queries.erase(it);
+}
+
+QueryMetricLogElement QueryMetricLog::createLogMetricElement(const String & query_id, std::shared_ptr<ProfileEvents::Counters::Snapshot> profile_counters, TimePoint current_time)
+{
+    std::lock_guard lock(queries_mutex);
+    auto query_status_it = queries.find(query_id);
+
+    QueryMetricLogElement elem;
+    elem.event_time = timeInSeconds(current_time);
+    elem.event_time_microseconds = timeInMicroseconds(current_time);
+    elem.query_id = query_status_it->first;
+    elem.memory = CurrentMetrics::values[CurrentMetrics::MemoryTracking];
+    elem.background_memory = CurrentMetrics::values[CurrentMetrics::MergesMutationsMemoryTracking];
```

Comment:
This seems odd. Are this values global (which wouldn't make sense to log as part of query metrics) or per thread (which then could not have `background_memory`) or per query somehow?

Related to this: I would expect to see `memory_usage` and `peak_memory_usage` from `QueryStatusInfo` as we do for query_log. Is there an issue with those values that leads us to use CurrentMetrics instead?

Reply:
This was indeed wrong. The `memory_usage` for the query and the `CurrentMetrics::MemoryTracking` was the same on the first implementation of this where I was collecting the data **on the same thread the query runs** (in the progressCallback hook). However, this is not the case anymore, so I was taking the memory value for the pool thread, which is very off.

Agree with you, I'm going to report just the `memory_usage` and `peak_memory_usage`. The former will be now the correct value for the query, and the `background_memory` doesn't make much sense.